File: src/common/library/corpus/profiling.ts
/* istanbul ignore file */

import {
  applyAndWithBitmasks,
  toBitMask,
} from "@/common/library/corpus/corpus_byte_utils";

const POW_2_24 = Math.pow(2, 24);

// Returns a seeded pseudo-random number generator (Mulberry32)
function mulberry32(seed: number): () => number {
  let t = seed >>> 0;
  return function () {
    t += 0x6d2b79f5;
    let r = Math.imul(t ^ (t >>> 15), 1 | t);
    r ^= r + Math.imul(r ^ (r >>> 7), 61 | r);
    return ((r ^ (r >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * Generates an array of unique random non-negative integers in increasing order.
 * @param count Number of integers to generate
 * @param upperBound Exclusive upper bound (max value is upperBound - 1)
 * @param seed Optional seed for reproducibility
 * @returns Sorted array of unique random integers
 */
export function uniqueRandomSortedInts(
  count: number,
  upperBound: number,
  seed?: number
): number[] {
  if (count > upperBound) {
    throw new Error("Count cannot be greater than upperBound.");
  }
  const rand = mulberry32(seed ?? Date.now());
  // Generate [0, 1, ..., upperBound-1]
  const arr = Array.from({ length: upperBound }, (_, i) => i);
  // Fisher-Yates shuffle up to count
  for (let i = 0; i < count; ++i) {
    const j = i + Math.floor(rand() * (upperBound - i));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  const result = arr.slice(0, count);
  result.sort((a, b) => a - b);
  return result;
}

interface DataArray {
  bitmask8: Uint8Array;
  bitmask: Uint32Array;
  indices: number[];
  upperBound: number;
  numElements: number;
}

function createDataArray(
  count: number,
  upperBound: number = POW_2_24,
  seed?: number
): DataArray {
  const indices = uniqueRandomSortedInts(count, upperBound, seed);
  const bitmask = toBitMask(indices, upperBound);
  const bitmask8 = new Uint8Array(bitmask.buffer);
  return { bitmask, bitmask8, indices, upperBound, numElements: count };
}

/**
 * Checks whether two Uint8Array instances have the same underlying value.
 * @param a First Uint8Array
 * @param b Second Uint8Array
 * @returns true if equal, false otherwise
 */
export function bitmaskEquals(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function createRandomDataArrays(
  upperBound: number,
  sizes: number[],
  seed = 42
): DataArray[] {
  return Array.from({ length: sizes.length }, (_, i) =>
    createDataArray(sizes[i], upperBound, seed + i)
  );
}

function printDataSummary(times: number[]) {
  function percentile(arr: number[], p: number): number {
    if (arr.length === 0) return NaN;
    const sorted = [...arr].sort((a, b) => a - b);
    const pos = (p / 100) * (sorted.length - 1);
    const lower = Math.floor(pos);
    const upper = Math.ceil(pos);
    const weight = pos - lower;
    return sorted[lower] * (1 - weight) + sorted[upper] * weight;
  }

  const mean = times.reduce((sum, t) => sum + t, 0) / times.length;
  const percentileKeys = [5, 10, 25, 50, 75, 90, 95];
  const percentileLabels = percentileKeys.map((p) => `p${p}`.padEnd(3, " "));
  const percentileValues = percentileKeys.map(
    (p) => percentile(times, p).toFixed(2) + " ms"
  );

  const sortedTimes = times.map((t, i) => ({ t, i })).sort((a, b) => a.t - b.t);

  const tail = 5;
  const tailSize = Math.ceil(times.length * (tail / 100));

  const minTail = sortedTimes.slice(0, tailSize);
  const maxTail = sortedTimes.slice(-tailSize);

  console.log(
    `Mean: ${mean.toFixed(2)} ms per iteration [${times.length} iterations]`
  );
  console.log("Percentiles:");
  for (let i = 0; i < percentileKeys.length; i++) {
    console.log(`- ${percentileLabels[i]}: ${percentileValues[i]}`);
  }
  console.log(
    `Min ${tail}%: ${minTail
      .map((x) => x.t.toFixed(2))
      .join(", ")} (indices: ${minTail.map((x) => x.i).join(", ")})`
  );
  console.log(
    `Max ${tail}%: ${maxTail
      .map((x) => x.t.toFixed(2))
      .join(", ")} (indices: ${maxTail.map((x) => x.i).join(", ")})`
  );
}

function runProfiling() {
  const fraction = 1024;
  const sizes = Array(4).fill(POW_2_24 / fraction);
  const data = createRandomDataArrays(POW_2_24, sizes, 42);

  let last: any = undefined;
  const times: number[] = [];
  const reps = 100;

  for (let i = 0; i < reps; i++) {
    const k = (i % 2) * 2;
    const a = data[k];
    const b = data[k + 1];

    const start = performance.now();
    // Add the operation to be measured below
    last = applyAndWithBitmasks(a.bitmask, b.bitmask, 3);
    // Add the operation to be measured above
    times.push(performance.now() - start);
  }

  printDataSummary(times);
  console.log(
    `${POW_2_24} bits (${POW_2_24 / fraction} set), ${last.length} words`
  );
}

runProfiling();

/*
To run:
  bun build --outfile=binarystuff.js --minify src/common/library/corpus/profiling.ts --format=iife \
    && node binarystuff.js && rm binarystuff.js
 */

File: src/rust/src/profiling.rs
mod bitmask_in_place_utils;
mod packed_index_utils;

use std::time::{Instant};
use packed_index_utils::smear_bitmask;
use packed_index_utils::to_bitmask;

const POW_2_24: u32 = 1 << 24;

struct Mulberry32 {
    t: u32,
}
impl Mulberry32 {
    fn new(seed: u32) -> Self {
        Mulberry32 { t: seed }
    }
    fn next(&mut self) -> f64 {
        self.t = self.t.wrapping_add(0x6d2b79f5);
        let mut r = (self.t ^ (self.t >> 15)).wrapping_mul(1 | self.t);
        r ^= r.wrapping_add((r ^ (r >> 7)).wrapping_mul(61 | r));
        ((r ^ (r >> 14)) as u64 as f64) / 4294967296.0
    }
}

/// Generates an array of unique random non-negative integers in increasing order.
fn unique_random_sorted_ints(count: u32, upper_bound: u32, seed: Option<u32>) -> Vec<u32> {
    assert!(count <= upper_bound, "Count cannot be greater than upperBound.");
    let mut rand = Mulberry32::new(seed.unwrap_or_else(|| {
        use std::time::SystemTime;
        SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs() as u32
    }));
    let mut arr: Vec<u32> = (0..upper_bound).collect();
    // Fisher-Yates shuffle up to count
    for i in 0..count {
        let j = i + ((rand.next() * ((upper_bound - i) as f64)) as u32);
        arr.swap(i as usize, j as usize);
    }
    let mut result = arr[..count as usize].to_vec();
    result.sort_unstable();
    result
}


struct DataArray {
    #[allow(dead_code)]
    bitmask8: Vec<u8>,
    bitmask: Vec<u64>,
    #[allow(dead_code)]
    indices: Vec<u32>,
    #[allow(dead_code)]
    upper_bound: u32,
    #[allow(dead_code)]
    num_elements: u32,
}

fn create_data_array(count: u32, upper_bound: u32, seed: Option<u32>) -> DataArray {
    let indices = unique_random_sorted_ints(count, upper_bound, seed);
    let bitmask = to_bitmask(&indices, upper_bound);
    let bitmask8 = bitmask.iter().flat_map(|w| w.to_le_bytes()).collect();
    DataArray {
        bitmask8,
        bitmask,
        indices,
        upper_bound,
        num_elements: count,
    }
}

fn create_random_data_arrays(upper_bound: u32, sizes: &[u32], seed: u32) -> Vec<DataArray> {
    sizes.iter().enumerate()
        .map(|(i, &sz)| create_data_array(sz, upper_bound, Some(seed + i as u32)))
        .collect()
}

fn percentile(arr: &[f64], p: f64) -> f64 {
    if arr.is_empty() { return f64::NAN; }
    let mut sorted = arr.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let pos = p / 100.0 * (sorted.len() as f64 - 1.0);
    let lower = pos.floor() as usize;
    let upper = pos.ceil() as usize;
    let weight = pos - lower as f64;
    sorted[lower] * (1.0 - weight) + sorted[upper] * weight
}

fn print_data_summary(times: &[f64]) {
    let mean: f64 = times.iter().sum::<f64>() / times.len() as f64;
    let percentile_keys = [5.0, 10.0, 25.0, 50.0, 75.0, 90.0, 95.0];
    println!("Mean: {:.2} ms per iteration [{} iterations]", mean, times.len());
    println!("Percentiles:");
    for &p in &percentile_keys {
        println!("- p{:2}: {:.2} ms", p as u32, percentile(times, p));
    }
    let mut sorted_times: Vec<(f64, usize)> = times.iter().copied().enumerate().map(|(i, t)| (t, i)).collect();
    sorted_times.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
    let tail = 5;
    let tail_size = ((times.len() as f64 * (tail as f64 / 100.0)).ceil()) as usize;
    let min_tail = &sorted_times[..tail_size];
    let max_tail = &sorted_times[sorted_times.len()-tail_size..];
    println!("Min {}%: {} (indices: {})",
        tail,
        min_tail.iter().map(|x| format!("{:.2}", x.0)).collect::<Vec<_>>().join(", "),
        min_tail.iter().map(|x| x.1.to_string()).collect::<Vec<_>>().join(", ")
    );
    println!("Max {}%: {} (indices: {})",
        tail,
        max_tail.iter().map(|x| format!("{:.2}", x.0)).collect::<Vec<_>>().join(", "),
        max_tail.iter().map(|x| x.1.to_string()).collect::<Vec<_>>().join(", ")
    );
}

fn run_profiling() {
    let fraction: u32 = 1024;
    let size = POW_2_24 / fraction;
    let sizes = vec![size; 4];
    let data = create_random_data_arrays(POW_2_24, &sizes, 42);

    let mut last_len = 0usize;
    let mut times = Vec::new();
    let reps = 100;

    for i in 0..reps {
        let k = (i % 2) * 2;
        let a = &data[k];
        // let b = &data[k + 1];

        let start = Instant::now();
        // Add the operation to be measured below
        smear_bitmask(&a.bitmask, 15, "both");
        // Add the operation to be measured above
        let elapsed = start.elapsed();
        last_len = a.bitmask.len();
        times.push(elapsed.as_secs_f64() * 1000.0);
    }

    print_data_summary(&times);
    println!(
        "{} bits ({} set), {} words",
        POW_2_24,
        POW_2_24 / fraction,
        last_len
    );
}

fn main() {
    run_profiling();
}

/*
To run:
  cargo run --bin profiling
*/
